dmoj/cptbox/isolate.py

```diff
@@ -314,63 +348,97 @@ def _file_access_check(
-                return real, ACCESS_EACCES
-
-        return normalized, None
-
-    def get_full_path(self, debugger: Debugger, file: str, dirfd: int = AT_FDCWD) -> str:
-        dirfd = (dirfd & 0x7FFFFFFF) - (dirfd & 0x80000000)
-        if not file.startswith('/'):
-            dir = self._getcwd_pid(debugger.tid) if dirfd == AT_FDCWD else self._getfd_pid(debugger.tid, dirfd)
-            file = os.path.join(dir, file)
-        file = '/' + os.path.normpath(file).lstrip('/')
-        return file
+                raise AccessFailure(ACCESS_EACCES, 'Denying %s, real path %s', file, real)
 
-    def do_kill(self, debugger: Debugger) -> bool:
+    def handle_kill(self, debugger: Debugger) -> None:
         # Allow tgkill to execute as long as the target thread group is the debugged process
         # libstdc++ seems to use this to signal itself, see <https://github.com/DMOJ/judge/issues/183>
-        return True if debugger.uarg0 == debugger.pid else ACCESS_EPERM(debugger)
+        if debugger.uarg0 != debugger.pid:
+            raise AccessFailure(ACCESS_EPERM, 'Cannot kill other processes.')
 
-    def do_prlimit(self, debugger: Debugger) -> bool:
-        return True if debugger.uarg0 in (0, debugger.pid) else ACCESS_EPERM(debugger)
+    def handle_prlimit(self, debugger: Debugger) -> None:
+        if debugger.uarg0 not in (0, debugger.pid):
+            raise AccessFailure(ACCESS_EPERM, 'Cannot prlimit other processes')
 
-    def do_prctl(self, debugger: Debugger) -> bool:
+    def handle_prctl(self, debugger: Debugger) -> None:
         PR_GET_DUMPABLE = 3
         PR_SET_NAME = 15
         PR_GET_NAME = 16
         PR_SET_THP_DISABLE = 41
         PR_SET_VMA = 0x53564D41  # Used on Android
-        return debugger.arg0 in (PR_GET_DUMPABLE, PR_SET_NAME, PR_GET_NAME, PR_SET_THP_DISABLE, PR_SET_VMA)
+        if debugger.arg0 not in (PR_GET_DUMPABLE, PR_SET_NAME, PR_GET_NAME, PR_SET_THP_DISABLE, PR_SET_VMA):
+            raise AccessFailure(protection_fault, 'Invalid prctl argument')
+
+    # ignore typing because of overload checks
+    def update(self, handlers: Mapping[int, Any]) -> None:  # type: ignore
+        for syscall, handler in handlers.items():
+            self[syscall] = handler
+
+    def __setitem__(self, syscall: int, handler) -> None:
+        if handler == ALLOW or isinstance(handler, ErrnoHandlerCallback):
+            super().__setitem__(syscall, handler)
+        else:
+            super().__setitem__(syscall, wrap_access_check(syscall, handler))
+
+
+def wrap_access_check(syscall: int, check: AccessChecker) -> HandlerCallback:
+    def inner(debugger) -> bool:
+        try:
+            check(debugger)
+            return True
+        except AccessFailure as failure:
+            failure.log(syscall)
+            return failure.handle(debugger)
+
+    return inner
+
+
+def protection_fault(self, debugger: Debugger) -> bool:
+    return False
+
+
+class AccessFailure(Exception):
```

Comment:
Nit: this is no longer just an "access" failure, as it's used for disallowing other syscalls. Probably `DeniedSyscall` or something would be better? Worth updating the log message too.